zkstore/store.go

```diff
@@ -0,0 +1,446 @@
+package zkstore
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/binary"
+	"hash"
+	"path"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/samuel/go-zookeeper/zk"
+)
+
+// IStore is the interface to which Store confirms. Documentation for these
+// methods lives on the concrete Store type, as the NewStore method returns
+// a concrete type, not this interface.  Clients may choose to use the IStore
+// interface if they wish.
+type IStore interface {
+	Put(item Item) (Ident, error)
+	Get(ident Ident) (item Item, found bool, err error)
+	List(category string) (locations []Location, found bool, err error)
+	Versions(location Location) (versions []string, found bool, err error)
+	Delete(ident Ident) (found bool, err error)
+	Close() error
+}
+
+// ensure that Store confirms to the IStore interface.
+var _ IStore = &Store{}
+
+// HashProviderFunc is a factory for hashers.
+type HashProviderFunc func() hash.Hash
+
+// Store exposes an API for performing CRUD operations against a backing
+// ZK cluster.
+type Store struct {
+	conn             *zk.Conn                  // the zk connection
+	basePath         string                    // the base path to use for any znodes
+	bucketsZnodeName string                    // the name of the znode folder
+	acls             []zk.ACL                  // the ACLs to use for any created nodes
+	hashBuckets      int                       // the number of hash buckets
+	bucketFunc       func(string) (int, error) // converts a name into a bucket number
```

Comment:
Converting a name into a bucket number requires the `hashBuckets` value, though this is accessible from the store in the `StoreOpt` context, why not making it an argument? `func(int, string) (int, error)`